### PR TITLE
fix #7955 fully specialized generic failed at array construction

### LINKED ISSUE #7955: fully specialized generic failed at array construction
```Nim
type
  SBase[T, V] = ref object of RootObj
    val: T
    color: V
  SRC = ref object of SBase[string, int]
  SRD = ref object of SBase[string, int]

var x = [a, b, c] #ok
var y = [b, c, a] #failed
var z = [c, b, a] #failed
```

what need to be included in the test:
 * the above case
 * secondary form of ref -> e.g. from new(ref someobject)
 * ptr form, and it's secondary form
 * ordinary object -> perhaps generate runtime error because of object slicing

see the evolution of this bug in #7601, #7818, and #7906 

note: if possible, uncomment test at `block test_t4799_6` of #4799

### PR DESCRIPTION
**Pull request by bung87**
fix #7955

related: #7956 (test case covered by t4799_2.nim t4799_3.nim )

### CODE REVIEW

**Comment by bung87** [comment]
ci fails summary:  
1. ` tests/typerel/t4799_2.nim`  `testVehicle([b, c, v])` fails , change to `testVehicle(b, c, v)` pass.

2. t4799.nim(24, 20) Error: type mismatch: got <GRD[system.int]> but expected 'GRBase:ObjectType = object'

**Comment by Araq** [comment]
Feel free to change the test case but beware that important packages keep working. :-)

**Comment by bung87** [comment]
> Feel free to change the test case but beware that important packages keep working. :-)

okay, I'll do that, also previous behavior assuming developer want fill element same as previous type, after this the error message should change, do you have suggesstion what's the message would like ?

File: compiler/types.nim [hunk 1]
```diff
@@ -1293,7 +1293,10 @@ proc commonSuperclass*(a, b: PType): PType =
     if ancestors.contains(y.id):
       # bug #7818, defer the previous skipTypes
       if t.kind != tyGenericInst: t = y
-      return t
+      if ancestors.contains(t.id):
```
Comment (Araq) [inline compiler/types.nim]: Please document your reasoning behind this logic.

File: tests/typerel/t4799_2.nim [hunk 2]
```diff
@@ -18,4 +18,4 @@ proc testVehicle[T](x: varargs[Vehicle[T]]): string =
 var v = Vehicle[int](tire: 3)
 var c = Car[int](tire: 4)
 var b = Bike[int](tire: 2)
-echo testVehicle([b, c, v])
\ No newline at end of file
+echo testVehicle([v, c, b])
```
Comment (Araq) [inline tests/typerel/t4799_2.nim]: Why was this necessary? The point of `commonType` is to compute the type to be `Vehicle[int]` regardless of the order in the array construction!

Comment (bung87) [reply tests/typerel/t4799_2.nim]: see https://github.com/nim-lang/Nim/issues/7955#issuecomment-1251817322

Comment (Araq) [reply tests/typerel/t4799_2.nim]: I don't agree with your comment.

Comment (bung87) [reply tests/typerel/t4799_2.nim]: I'd say the issue case (`var y = [b, c] #failed`) is the worse case, neither previous nor current implementation fullfill your statement.

Comment (bung87) [reply tests/typerel/t4799_2.nim]: to be clear, check `proc semArrayConstr` the implementation says order are matters. and I remenber `typeRel` only has `subType` to describe this, there's no things like `commonType` or `fatherType`.

Comment (Araq) [reply tests/typerel/t4799_2.nim]:
```
nimgrep commonType --ext:nim compiler/
compiler/sem.nim:126: template commonTypeBegin*(): PType = PType(kind: tyUntyped)
compiler/sem.nim:128: proc commonType*(c: PContext; x, y: PType): PType =
compiler/sem.nim:216: proc commonType*(c: PContext; x: PType, y: PNode): PType =
compiler/sem.nim:219:   commonType(c, x, y.typ)
compiler/semexprs.nim:660:       typ = commonType(c, typ, xx.typ)
compiler/semexprs.nim:2423:   var typ = commonTypeBegin
compiler/semexprs.nim:2441:           typ = commonType(c, typ, it[1].typ)
compiler/semexprs.nim:2457:           typ = commonType(c, typ, it[0].typ)
compiler/semstmts.nim:160:   var typ = commonTypeBegin
compiler/semstmts.nim:169:       typ = commonType(c, typ, it[1])
compiler/semstmts.nim:175:       typ = commonType(c, typ, it[0])
compiler/semstmts.nim:211:   var typ = commonTypeBegin
compiler/semstmts.nim:214:   typ = commonType(c, typ, n[0].typ)
compiler/semstmts.nim:274:       typ = commonType(c, typ, a[^1])
compiler/semstmts.nim:1075:   var typ = commonTypeBegin
compiler/semstmts.nim:1109:       typ = commonType(c, typ, x[last])
compiler/semstmts.nim:1117:       typ = commonType(c, typ, x[1])
compiler/semstmts.nim:1123:       typ = commonType(c, typ, x[0])
```

Comment (bung87) [reply tests/typerel/t4799_2.nim]: hmm , I get it

**Comment by bung87** [comment]
found `commonType` just works fine, the problem is when the `result[i]` goes to fitNode -> paramTypesMatchAux -> typeRel, it doesn't count commonSuperClass as `commonType` did.

File: compiler/ast.nim [hunk 3]
```diff
@@ -583,6 +583,7 @@ type
     tfIsConstructor
     tfEffectSystemWorkaround
     tfIsOutParam
+    tfCommonType
```
Comment (Araq) [inline compiler/ast.nim:586]: No.

File: compiler/semexprs.nim [hunk 4]
```diff
@@ -662,6 +662,8 @@ proc semArrayConstr(c: PContext, n: PNode, flags: TExprFlags; expectedType: PTyp
       #result.add fitNode(c, typ, n[i])
       inc(lastIndex)
     addSonSkipIntLit(result.typ, typ, c.idgen)
+    if typ.skipTypes({tyPtr, tyRef}).kind in {tyGenericInst, tyObject}:
```
Comment (Araq) [inline compiler/semexprs.nim:665]: An object is a concrete nominal type and a tyGenericInst is an instance of a generic, not necessarily a generic from an object so this check cannot possibly be correct, no matter why it exists.

**Comment by Araq** [comment]
Sorry but that's just hacking stuff together until the test case works and nothing else breaks. What is a `tfCommonType` flag? It's certainly not a property of the type system.

**Comment by bung87** [comment]
the `semArrayConstr` query the common base type of elements and take it as element type. if this patch does not work as expected then I assuming the test case is invalid.

File: compiler/semexprs.nim (hunk 4 above)
Comment (bung87) [reply compiler/semexprs.nim:665]: here is just ensure it gonna compare object types for array construction in `typeRel`, maybe should use some like `skipToObject`.

**Comment by bung87** [comment]
succeed by https://github.com/nim-lang/Nim/pull/20612
